src/architecture/unixfs.md

```diff
@@ -0,0 +1,618 @@
+
+This allows to do fast indexing into the file, 	for example if someone is trying
+to read bytes 25 to 35 we can compute an offset list by summing all previous
+indexes in `blocksizes`, then do a search to find which indexes contain the
+range we are interested in.
+
+For example here the offset list would be `[0, 20]` and thus we know we only need to download `Qmbar` to get the range we are intrested in.
+
+UnixFS parser MUST error if `blocksizes` or `Links` are not of the same length.
+
+##### `decode(PBNode.Data).Data`
+
+This field is an array of bytes, it is the file content and is appended before
+the links. This must be taken into account when doing offset calculations, that is
+the length of `decode(PBNode.Data).Data` defines the value of the zeroth element
+of the offset list when computing offsets.
+
+##### `PBNode.Links[].Name`
+
+This field makes sense only in :ref[Directories] contexts and MUST be absent
+when creating a new file. For historical reasons, implementations parsing
+third-party data SHOULD accept empty values here.
+
+If this field is present and non-empty, the file is invalid and the parser MUST
+error.
+
+##### `decode(PBNode.Data).Blocksize`
+
+This field is not directly present in the block, but rather a computable property
+of a `dag-pb`, which would be used in the parent node in `decode(PBNode.Data).blocksizes`.
+It is the sum of the length of `decode(PBNode.Data).Data` field plus the sum
+of all link's `blocksizes`.
+
+##### `decode(PBNode.Data).filesize`
+
+If present, this field MUST be equal to the `Blocksize` computation above.
+Otherwise, this file is invalid.
+
+##### Path Resolution
+
+A file terminates a UnixFS content path. Any attempt to resolve a path past a
+file MUST error.
+
+#### `Directory` Type
+
+A :dfn[Directory], also known as folder, is a named collection of child :ref[Nodes]:
+
+- Every link in `PBNode.Links` is an entry (child) of the directory, and
+  `PBNode.Links[].Name` gives you the name of such child.
+- Duplicate names are not allowed. Therefore, two elements of `PBNode.Link` CANNOT
+  have the same `Name`. if two identical names are present in a directory, the
+  decoder MUST fail.
+
+The minimum valid `PBNode.Data` field for a directory is as follows:
+
+```json
+{
+  "Type": "Directory"
+}
+```
+
+The remaining relevant values are covered in [Metadata](#metadata).
+
+##### Link Ordering
+
+The canonical sorting order is lexicographical over the names.
+
+In theory there is no reason an encoder couldn't use an other ordering, however
+this lose some of its meaning when mapped into most file systems today (most file
+systems consider directories are unordered-key-value objects).
+
+A decoder SHOULD, if it can, preserve the order of the original files in however
+it consume those names. However when, some implementation decode, modify then
+re-encode some, the original links order fully lose it's meaning (given that there
+is no way to indicate which sorting was used originally).
```

Comment:
```suggestion
In theory, there is no reason an encoder couldn't use an other ordering. However,
this loses some of its meaning when mapped into most file systems today, as most file
systems consider directories to be unordered key-value objects.

A decoder SHOULD, if it can, preserve the order of the original files in the same way
it consumed those names. However, when some implementations decode, modify and then
re-encode, the original link order loses it's original meaning, given that there
is no way to indicate which sorting was used originally.
```